src/core/processing/models/qgsprocessingmodelalgorithm.cpp

```diff
@@ -530,6 +531,9 @@ QMap<QString, QgsProcessingModelAlgorithm::VariableDefinition> QgsProcessingMode
     if ( !layer )
       layer = QgsProcessingUtils::mapLayerFromString( value.toString(), context );
 
+    // TODO: should I tag variable name with type? e.g. %1_maplayer
+    variables.insert( safeName( QStringLiteral( "%1" ).arg( name ) ), VariableDefinition( value, source, description ) );
```

Comment:
This should possibly be:

    VariableDefinition( QObject::fromValue( layer ),... )

Reply:
I suppose QVariant instead of QObject